### ISSUE TITLE
[BOUNTY] Waku Meme Board 2.0

### ISSUE THREAD

**Issue by fryorcraken** [waku-org/bounties#16]
## Context

Waku Is Uncompromising Web3 Communication at Scale.
A family of robust, censorship-resistant communication protocols, designed to enable privacy-focused messaging for web3 apps.

The JavaScript implementation of Waku, [js-waku](https://github.com/waku-org/js-waku/), enables web apps to utilize the Waku network for off-chain message transmission.

Disclaimer: Waku is **experimental**, you may find blocking issues while developing your web app. We will prioritize their resolution to unblock you, which means you may have to pause development until done.
Thank you for your patience and understanding.

**Your participation in this bounty is subject to your acceptance of our terms and conditions.**
Please see https://github.com/waku-org/bounties#applying-for-a-bounty for details.

## Reward

USD1000 in DAI.

## Timeframe to Completion

Once the application is approved, the result must be submitted within 30 days.

## Application Evaluation

To ensure you are selected for this bounty, provide the following information:

- Describe the approach you will take to tackle the bounty.
- A timeline within which you expect to deliver the bounty.
- Any previous relevant work you have done.

## Bounty

The Waku Meme board is a dApp that uses Waku and BitTorrent to display a timeline of memes.
Post memes on the board is permission-less, so is viewing memes.

### Impact

The aim of the bounty is to demonstrate how Waku can be used for ephemeral communication while offloading large files to another, more appropriate, system such as BitTorrent.

This will hopefully provide the inspiration to developers to combine Waku with other decentralized technology such as BitTorrent.

### Deliverables

A Web dApp that enables the following:
- Display previously posted memes (latest first)
- Post memes (images only, no comment/text).

The dApp is expected to be designed as follows:

- A unique content topic is used for the dApp, such as `/meme-board/2/meme/proto`
- When posting a meme, the dApp upload the image to BitTorrent and then send a Waku message with Waku Light push that contains:
    - magnet link to the image
    - Protobuf must be used for payload encoding
- When opening the dApp, and loading the timeline:
   - Waku Messages are retrieved via store + filter on the unique content topic
   - Images are retrieved from BitTorrent using the info in the Waku messages
   - Memes are displayed, newest first and on top.

It is recommended to use https://github.com/webtorrent/webtorrent for the BitTorrent integration.

Please check with the Waku team if you decide to use a different design.
Waku **must** be used for communication purposes.

### Acceptance Criteria

- The dApp works as described:
	- memes are displayed latest on top
	    - Some BitTorrent meta info to be displayed near each meme such as seed data (upload) and # of connected peers
	- Waku Store is used to retrieved older memes
	- Waku Filter is used for real-time reception of memes
	- No image data is uploaded to Waku, only BitTorrent magnet links
	- Some visual feedback is provided to the user when uploading the image:
	    - Number of BitTorrent peers connected to
	    - Chunk upload status
	- Users can send memes.
- Must use the latest stable version of `@waku/sdk`
- Must be a frontend only, browser web app (no backend, no NodeJS, no SSR, no lambda function otherwise it's not decentralized). 
- Host result GitHub pages or equivalent
- All code must be licensed under MIT+Apache2.0.

### Resources

Learn more about Waku at https://docs.waku.org/.
Join our Discord to get support at https://discord.waku.org/.
Various examples available at https://github.com/waku-org/js-waku-examples/.

**Comment by b4s36t4** [comment]
@fryorcraken Sorry for my enthusiasm but I'm just wondering about the difference between v1 and v2? it's just how we upload the memes right? on V1 we did used IPFS for this bounty we have to use webtorrent right? 

> Waiting for this bounty to be opened 🤓

**Comment by fryorcraken** [comment]
> @fryorcraken Sorry for my enthusiasm but I'm just wondering about the difference between v1 and v2? it's just how we upload the memes right? on V1 we did used IPFS for this bounty we have to use webtorrent right?
> 
> > Waiting for this bounty to be opened 🤓

We learned that IPFS was not ideal so let's try with BitTorrent

**Comment by b4s36t4** [comment]
Cool, I'd love to propose my solution when it opens.

**Comment by hackyguru** [comment]
@b4s36t4 another suggestion from my side would be to use a pinning service like [web3.storage](https://web3.storage) or [nft.storage](https://nft.storage) to avoid IPFS note setup complications

**Comment by b4s36t4** [comment]
@hackyguru yes at first I did the same, used pinata only. But since we're trying to achieve full decentralization using a ipfs service is not an option to go.

**Comment by fryorcraken** [comment]
@hackyguru what'st the decentralization level of these services?

@b4s36t4 from nft.storage:

> Free decentralized storage and bandwidth for NFTs on IPFS and Filecoin.

**Comment by b4s36t4** [comment]
I think nft.storage and web3.storage is from same. People (protocol labs) but they require a user to login and create an API then using their sdks to upload the assets. Which is same as pinata. These all services uses ipfs under the hood. 

What about lighthouse.storage? @fryorcraken

**Comment by fryorcraken** [comment]
Thanks for this. Not sure about lighthouse but either way, keen to get some input from @hackyguru

**Comment by vpavlin** [comment]
> Same thing - they all use API gateway. The underlying storage might be decentralized, but the web interaction is via REST API. Sadly IPFS does not seem to have an answer to this at the moment (helia is useless for providing content).

WebTorrent which relies on WebRTC, thus can potentially allow direct connection between browser instances, and thus minimize the 3rd party involvement, would make sense to me. Although I am not entirely how it will work in context of Memeboard content - since when someone posts and closes the window, no one is seeding the content anymore, right? So it might needs some kind of backend actually getting the initial seed and then being the seeder once the window is closed?

**Comment by fryorcraken** [comment]
> . Although I am not entirely how it will work in context of Memeboard content - since when someone posts and closes the window, no one is seeding the content anymore, right? So it might needs some kind of backend actually getting the initial seed and then being the seeder once the window is closed?

Meme poster should keep their window open.

We can then do a follow-up bounty with go-waku or nwaku REST API to have a backend that seed memes.

**Comment by b4s36t4** [comment]
Well, I think we can use FVM to support files & keep them seeding. As Filecoin is full decentralized and we don't need to update anything to keep the files available it should fair play using filecoin and FVM.

> I don't know FVM fully :)

**Comment by fryorcraken** [comment]
> Well, I think we can use FVM to support files & keep them seeding. As Filecoin is full decentralized and we don't need to update anything to keep the files available it should fair play using filecoin and FVM.
> 
> > I don't know FVM fully :)

Let's try bittorrent this time and we can have future bounty with fvm

**Comment by b4s36t4** [comment]
Understood.

**Comment by hackyguru** [comment]
All pinning services including lighthouse.storage are centralised at some point (obviously because they have API Keys). But unfortunately native IPFS integration has a lot of problems in terms of file availability :( Let's go ahead with Bittorrent, excited to see the outcome!!!!

**Comment by b4s36t4** [comment]
@hackyguru, is this up for grabs? if so shall I proceed and submit a proposal :)

**Comment by fryorcraken** [comment]
We want to make RLN by defualt on js-waku first before releasing more js-waku bounties.
